### PR TITLE
#63 Fix/add to cart

### PR DESCRIPTION
**Pull request by minhng271**
### Fix bug feature add to cart
- Each item is only added once
- The number should allow to update by typing
- Quantity 0 does not allow adding to cart
### Result

https://user-images.githubusercontent.com/49194444/215661844-d36f6a89-9926-4aee-970a-3b1aee8ff416.mp4

### Update Result

https://user-images.githubusercontent.com/49194444/215941602-e117610d-d6f1-4783-8d10-fe470936d333.mp4

### CODE REVIEW

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 1]
```diff
@@ -59,17 +59,25 @@ const ProductDetail = () => {
...
-          ...listItem.listProducts,
-          {
-            idProduct: product.id,
-            quantity: quantityProduct,
-            name: product.name,
-            price: product.price,
-          }],
-      };
+      let cartUser = {};
+      const c = listItem?.listProducts?.findIndex((item) => item.idProduct === Number(productId));
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: Do not name variables like this: a, b, c.... Variable must have a clear meaning.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 2]
```diff
@@ -59,17 +59,25 @@ const ProductDetail = () => {
...
+      if (c > 0) {
+        listItem.listProducts[c].quantity += quantityProduct;
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: ` const { listItem, setListItem } = useCart();` => `const { cart, setCart } = useCart()`

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 3]
```diff
@@ -59,17 +59,25 @@ const ProductDetail = () => {
...
+        cartUser = listItem;
+      } else {
+        cartUser = {
+          id: userData.userId,
+          listProducts: [
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: Just `products`.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 4]
```diff
@@ -94,68 +102,49 @@ const ProductDetail = () => {
...
-        <Toast
-          status={toast.status}
-          message={toast.message}
-          onClose={handleClose}
-        />
-        )
-      }
+          </div>
+          <div className="product-description">
+            <div className="title">Description</div>
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: You forgot basic HTML knowledge. Don't use div for text.

File: react-basic-practice/src/components/Quantity/index.jsx [hunk 5]
```diff
@@ -18,11 +19,15 @@ const Quantity = (props) => {
...
+    if (counter > 1) {
       setCounter((preCounter) => preCounter - 1);
     }
   };
 
+  const setQuantity = (value) => {
+    setCounter(value.quantity);
+  };
+
   if (!status) {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/Quantity/index.jsx]: Please explain this.

Comment (minhng271) [reply react-basic-practice/src/components/Quantity/index.jsx]: `status` is the status of the existing product, if it is `sold out`, the data will not be obtained from the quantity button, if the status is false, the value can be obtained from the quantity button

Comment (MinhNguyen41092) [reply react-basic-practice/src/components/Quantity/index.jsx]: You are mixing data types. If it's string then all values must be string, you should never have many data types for the same prop or variable.

File: react-basic-practice/src/components/Quantity/index.jsx [hunk 6]
```diff
@@ -18,11 +19,15 @@ const Quantity = (props) => {
...
     quantity(counter);
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/Quantity/index.jsx]: A function/method must start with a verb.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 7]
```diff
@@ -59,19 +59,29 @@ const ProductDetail = () => {
...
-          ...listItem.listProducts,
-          {
-            idProduct: product.id,
-            quantity: quantityProduct,
-            name: product.name,
-            price: product.price,
-          }],
-      };
+      let cartUser = {};
+      const haveProduct = cart?.products?.findIndex(
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: Use `array.some` instead of `findIndex`.

Comment (minhng271) [reply react-basic-practice/src/pages/ProductDetail/index.jsx]: I need the position of the item to update in the array so use `findIndex`

Comment (MinhNguyen41092) [reply react-basic-practice/src/pages/ProductDetail/index.jsx]: Then you must rename the variable.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 8]
```diff
@@ -26,6 +26,7 @@ const ProductDetail = () => {
   const { loading, setLoading } = useLoading();
   const [product, setProduct] = useState({});
   const [quantityProduct, setQuantity] = useState(0);
+  const [unavailableProduct, setUnavailableProduct] = useState(false);
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: Redundant if you don't check product availability in the database.

File: react-basic-practice/src/pages/ProductDetail/index.jsx [hunk 9]
```diff
@@ -92,70 +103,51 @@ const ProductDetail = () => {
     setToast({ ...toast, openPopup: false });
   };
 
+  const checkUnavailableProduct = () => {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/pages/ProductDetail/index.jsx]: reduce this to one line.

**Pull request merged**